File: 4000/4179.cpp
#include <iostream>
#include <queue>
using namespace std;
#define p pair<int, int>
#define MAX 1001

int r, c, answer = 1e9;
int dx[]{ 0,1,0,-1 };
int dy[]{ 1,0,-1,0 };
char area[MAX][MAX];
int ePath[MAX][MAX];
int fPath[MAX][MAX];
queue<p> q, q2;

void input() {
	cin >> r >> c;
	string s;
	for (int i = 0; i < r; ++i) {
		fill(ePath[i], ePath[i] + c, -1);
		fill(fPath[i], fPath[i] + c, -1);
	}
	for (int i = 0; i < r; ++i) {
		cin >> s;
		for (int j = 0; j < c; ++j) {
			area[i][j] = s[j];
			if (s[j] == 'J') {
				q2.push(p(i, j));
				ePath[i][j] = 0;
			}
			else if (s[j] == 'F') {
				q.push(p(i, j));
				fPath[i][j] = 0;
			}
		}
	}
}

void firePath() {
	while (!q.empty()) {
		p cur = q.front(); q.pop();
		for (int dir = 0; dir < 4; ++dir) {
			int nx = cur.first + dx[dir];
			int ny = cur.second + dy[dir];
			if (nx < 0 || ny < 0 || nx >= r || ny >= c) continue;
			if (fPath[nx][ny] >= 0 || area[nx][ny] == '#') continue;
			q.push(p(nx, ny));
			fPath[nx][ny] = fPath[cur.first][cur.second] + 1;
		}
	}
}

bool escapePath() {
	while (!q2.empty()) {
		p cur = q2.front(); q2.pop();
		for (int dir = 0; dir < 4; ++dir) {
			int nx = cur.first + dx[dir];
			int ny = cur.second + dy[dir];
			if (nx < 0 || ny < 0 || nx >= r || ny >= c) {
				cout << ePath[cur.first][cur.second] + 1;
				return 1;
			}
			if (ePath[nx][ny] >= 0 || area[nx][ny] == '#') continue;
			if (fPath[nx][ny] != -1 && fPath[nx][ny] <= ePath[cur.first][cur.second] + 1) continue;
			q2.push(p(nx, ny));
			ePath[nx][ny] = ePath[cur.first][cur.second] + 1;
		}
	}
	return 0;
}

void solution() {
	input();
	firePath();
	if (!escapePath()) cout << "IMPOSSIBLE";
}

int main(){
	cin.tie(0), cout.tie(0);
	ios_base::sync_with_stdio(0);
	solution();
	return 0;
}

File: 4000/4179.java
import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int n, m, answer;
	static char[][] area;
	static int[][] fire, path;
	static class Info{
		int x, y;
		Info(int x, int y){
			this.x = x;
			this.y = y;
		}
	}
	static Info init;
	static List<Info> fires;
	static final int[] dx = {0, 1, 0, -1};
	static final int[] dy = {1, 0, -1, 0};
	
    static void input() throws IOException {
		st = new StringTokenizer(br.readLine());
    	n = Integer.parseInt(st.nextToken());
    	m = Integer.parseInt(st.nextToken());
    	fires = new ArrayList<>();
    	area = new char[n][m];
    	fire = new int[n][m];
    	path = new int[n][m];
    	for(int i = 0;i < n; ++i) {
    		String tmp =br.readLine();
    		for(int j = 0;j < m; ++j) {
    			area[i][j] = tmp.charAt(j);
    			if(area[i][j] == 'J') init = new Info(i, j);
    			else if(area[i][j] == 'F') fires.add(new Info(i, j));
    		}
    	}
    	answer = (int)1e9;
    }
    
    static boolean range(int x, int y){
    	return x >= 0 && y >= 0 && x < n && y < m;
    }
    
    static void spread() {
    	Deque<Info> q = new ArrayDeque<>();
    	for(int i = 0;i < n; ++i) {
    		Arrays.fill(fire[i], (int)1e9);
    	}
    	for(Info info : fires) {
    		q.add(info);
    		fire[info.x][info.y] = 0; 
    	}
    	while(!q.isEmpty()) {
    		Info cur = q.poll();
    		for(int dir = 0;dir < 4; ++dir) {
    			int nx = cur.x + dx[dir];
    			int ny = cur.y + dy[dir];
    			if(!range(nx, ny) || area[nx][ny] == '#') continue;
    			if(fire[nx][ny] <= fire[cur.x][cur.y] + 1) continue;
    			fire[nx][ny] = fire[cur.x][cur.y] + 1; 
    			q.add(new Info(nx, ny));
    		}
    	}
    }
    
    static void escape() {
    	Deque<Info> q = new ArrayDeque<>();
    	for(int i = 0;i < n; ++i) {
    		Arrays.fill(path[i], (int)1e9);
    	}
    	q.add(init);
    	path[init.x][init.y] = 0; 
    	while(!q.isEmpty()) {
    		Info cur = q.poll();
    		for(int dir = 0;dir < 4; ++dir) {
    			int nx = cur.x + dx[dir];
    			int ny = cur.y + dy[dir];
    			if(!range(nx, ny)) {
    				answer = path[cur.x][cur.y] + 1;
    				return;
    			}
    			if(area[nx][ny] == '#') continue;
    			if(fire[nx][ny] <= path[cur.x][cur.y] + 1) continue;
    			if(path[nx][ny] <= path[cur.x][cur.y] + 1) continue;
    			path[nx][ny] = path[cur.x][cur.y] + 1; 
    			q.add(new Info(nx, ny));
    		}
    	}
    }
    
    static void solution() {
    	spread();
    	escape();
    	if(answer != (int)1e9) System.out.println(answer);
    	else System.out.println("IMPOSSIBLE");
    }
    
    public static void main(String[] args) throws IOException {
   		input();
   		solution();
    }
}
